# Model + Define, Train & Study

This stage focuses on selecting suitable regression algorithms for the **Repository Popularity Prediction** problem.

The target variable `stars` is a continuous numerical variable representing repository popularity.

The selected regression models will be loaded using Python and trained using the saved P2 feature-engineered training data from the Feature Engineering stage.

In [2]:
# Import libraries and create a connection to the AIDev MySQL database

import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://root:root@localhost:3306/aidev_project",
    pool_pre_ping=True,
    connect_args={
        "connect_timeout": 60,
        "read_timeout": 600,
        "write_timeout": 600
    }
)

print("Database connection created successfully.")

Database connection created successfully.


In [3]:
# Load P2 feature-engineered training and testing data

from scipy.sparse import load_npz
import numpy as np

# Load feature matrices
X_train = load_npz("../features/X_p2_train_final.npz")
X_test = load_npz("../features/X_p2_test_final.npz")

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X_train shape: (261438, 416751)
X_test shape: (65360, 416751)


In [4]:
# Load P2 target variable: stars
# Use row_order to guarantee alignment with the feature matrices.

query = """
SELECT
    row_order,
    split,
    stars
FROM ml_p2_split
ORDER BY row_order
"""

df_p2_split = pd.read_sql(
    query,
    engine
)

print("P2 target data loaded successfully.")
print("P2 target shape:", df_p2_split.shape)

print("\nColumns:")
print(df_p2_split.columns.tolist())

print("\nFirst 10 row_order values:")
print(df_p2_split["row_order"].head(10).tolist())

print("\nLast 10 row_order values:")
print(df_p2_split["row_order"].tail(10).tolist())

print("\nStars summary:")
print(df_p2_split["stars"].describe())

P2 target data loaded successfully.
P2 target shape: (326798, 3)

Columns:
['row_order', 'split', 'stars']

First 10 row_order values:
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Last 10 row_order values:
[326788, 326789, 326790, 326791, 326792, 326793, 326794, 326795, 326796, 326797]

Stars summary:
count    326798.000000
mean         89.202253
std        1978.829270
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max      375115.000000
Name: stars, dtype: float64


In [5]:
# Separate target variable into train and test sets
# The data is already ordered by row_order.

y_train = df_p2_split.loc[
    df_p2_split["split"] == "train", "stars"
].reset_index(drop=True)

y_test = df_p2_split.loc[
    df_p2_split["split"] == "test", "stars"
].reset_index(drop=True)

print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

print("\nTraining target summary:")
print(y_train.describe())

print("\nTesting target summary:")
print(y_test.describe())

y_train shape: (261438,)
y_test shape: (65360,)

Training target summary:
count    261438.000000
mean         89.592500
std        2036.206578
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max      375115.000000
Name: stars, dtype: float64

Testing target summary:
count     65360.000000
mean         87.641279
std        1730.413487
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max      135432.000000
Name: stars, dtype: float64


In [6]:
# Validate P2 feature and target alignment

print("P2 DATA VALIDATION")

print("X_train rows:", X_train.shape[0])
print("y_train rows:", len(y_train))

print("X_test rows :", X_test.shape[0])
print("y_test rows :", len(y_test))

print("\nFeature count:", X_train.shape[1])

print("\nMissing values in y_train:", y_train.isna().sum())
print("Missing values in y_test :", y_test.isna().sum())

print("\nAlignment check:")
print("Train:", X_train.shape[0] == len(y_train))
print("Test :", X_test.shape[0] == len(y_test))

print("\nRow order validation:")
print("Total target rows:", len(df_p2_split))
print("Expected rows:", X_train.shape[0] + X_test.shape[0])
print(
    "Row order continuous:",
    df_p2_split["row_order"].equals(
        pd.Series(range(len(df_p2_split)))
    )
)

P2 DATA VALIDATION
X_train rows: 261438
y_train rows: 261438
X_test rows : 65360
y_test rows : 65360

Feature count: 416751

Missing values in y_train: 0
Missing values in y_test : 0

Alignment check:
Train: True
Test : True

Row order validation:
Total target rows: 326798
Expected rows: 326798
Row order continuous: True


In [7]:
# Import regression algorithms

from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.preprocessing import PolynomialFeatures
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

print("All regression algorithms imported successfully.")

All regression algorithms imported successfully.


In [8]:
# Create a representative sample for regression model study

from sklearn.model_selection import train_test_split

sample_train_size = 50000
sample_test_size = 10000

X_train_sample, _, y_train_sample, _ = train_test_split(
    X_train,
    y_train,
    train_size=sample_train_size,
    random_state=42
)

X_test_sample, _, y_test_sample, _ = train_test_split(
    X_test,
    y_test,
    train_size=sample_test_size,
    random_state=42
)

print("Regression model-study sample created.")

print("\nX_train_sample:", X_train_sample.shape)
print("y_train_sample:", y_train_sample.shape)

print("X_test_sample :", X_test_sample.shape)
print("y_test_sample :", y_test_sample.shape)

print("\nTraining target summary:")
print(y_train_sample.describe())

print("\nTesting target summary:")
print(y_test_sample.describe())

Regression model-study sample created.

X_train_sample: (50000, 416751)
y_train_sample: (50000,)
X_test_sample : (10000, 416751)
y_test_sample : (10000,)

Training target summary:
count     50000.000000
mean         92.915980
std        2402.070736
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max      375115.000000
Name: stars, dtype: float64

Testing target summary:
count     10000.00000
mean        106.10670
std        2075.67121
min           0.00000
25%           0.00000
50%           0.00000
75%           0.00000
max      135432.00000
Name: stars, dtype: float64


In [9]:
# Define regression models

models = {
    "Linear Regression": LinearRegression(n_jobs=-1),

    "Lasso": Lasso(
        alpha=0.01,
        max_iter=1000
    ),

    "Ridge": Ridge(
        alpha=1.0
    ),

    "KNN": KNeighborsRegressor(
        n_neighbors=5,
        n_jobs=-1
    ),

    "SVM": SVR(
        kernel="linear"
    ),

    "Decision Tree": DecisionTreeRegressor(
        random_state=42
    ),

    "Random Forest": RandomForestRegressor(
        n_estimators=100,
        random_state=42,
        n_jobs=-1
    ),

    "XGBoost": XGBRegressor(
        n_estimators=100,
        max_depth=6,
        learning_rate=0.1,
        objective="reg:squarederror",
        eval_metric="rmse",
        random_state=42,
        n_jobs=-1
    )
}

print("Regression models defined successfully.")
print("\nModels:")
for model_name in models:
    print("-", model_name)

Regression models defined successfully.

Models:
- Linear Regression
- Lasso
- Ridge
- KNN
- SVM
- Decision Tree
- Random Forest
- XGBoost


---

## Linear Regression

Linear Regression is used as the baseline model for predicting repository popularity based on the engineered features.

In [10]:
# Train Linear Regression

linear_model = models["Linear Regression"]

linear_model.fit(
    X_train_sample,
    y_train_sample
)

print("Linear Regression trained successfully.")

Linear Regression trained successfully.


In [11]:
# Make predictions using Linear Regression

y_train_pred_lr = linear_model.predict(X_train_sample)
y_test_pred_lr = linear_model.predict(X_test_sample)

print("Linear Regression predictions completed.")
print("Training predictions:", len(y_train_pred_lr))
print("Testing predictions :", len(y_test_pred_lr))

Linear Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [12]:
# Evaluate Linear Regression

from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

train_rmse_lr = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_lr)
)

test_rmse_lr = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_lr)
)

train_r2_lr = r2_score(
    y_train_sample,
    y_train_pred_lr
)

test_r2_lr = r2_score(
    y_test_sample,
    y_test_pred_lr
)

print("Linear Regression Evaluation")
print("----------------------------")
print("Train RMSE:", train_rmse_lr)
print("Test RMSE :", test_rmse_lr)
print("Train R²  :", train_r2_lr)
print("Test R²   :", test_r2_lr)

Linear Regression Evaluation
----------------------------
Train RMSE: 7.860871492949246
Test RMSE : 1840.6020754090223
Train R²  : 0.9999892711639404
Test R²   : 0.21359515190124512


---

## Polynomial Regression

Polynomial Regression is used to capture non-linear relationships between the numerical repository features and repository popularity.

Due to the high-dimensional sparse feature matrix, Polynomial Regression is applied to the numerical feature subset to avoid excessive feature expansion.

In [13]:
# Train Polynomial Regression

from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

# The last 2 columns are the scaled numerical features
X_train_poly_base = X_train_sample[:, -2:].toarray()
X_test_poly_base = X_test_sample[:, -2:].toarray()

poly_features = PolynomialFeatures(
    degree=2,
    include_bias=False
)

X_train_poly = poly_features.fit_transform(X_train_poly_base)
X_test_poly = poly_features.transform(X_test_poly_base)

poly_model = LinearRegression()

poly_model.fit(
    X_train_poly,
    y_train_sample
)

print("Polynomial Regression trained successfully.")
print("Polynomial training features:", X_train_poly.shape)
print("Polynomial testing features :", X_test_poly.shape)

Polynomial Regression trained successfully.
Polynomial training features: (50000, 5)
Polynomial testing features : (10000, 5)


In [14]:
# Make predictions using Polynomial Regression

y_train_pred_poly = poly_model.predict(X_train_poly)
y_test_pred_poly = poly_model.predict(X_test_poly)

print("Polynomial Regression predictions completed.")
print("Training predictions:", len(y_train_pred_poly))
print("Testing predictions :", len(y_test_pred_poly))

Polynomial Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [15]:
# Evaluate Polynomial Regression

train_rmse_poly = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_poly)
)

test_rmse_poly = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_poly)
)

train_r2_poly = r2_score(
    y_train_sample,
    y_train_pred_poly
)

test_r2_poly = r2_score(
    y_test_sample,
    y_test_pred_poly
)

print("Polynomial Regression Evaluation")
print("---------------------------------")
print("Train RMSE:", train_rmse_poly)
print("Test RMSE :", test_rmse_poly)
print("Train R²  :", train_r2_poly)
print("Test R²   :", test_r2_poly)

Polynomial Regression Evaluation
---------------------------------
Train RMSE: 1037.0033751150474
Test RMSE : 1784.724558580399
Train R²  : 0.8136208057403564
Test R²   : 0.2606181502342224


---

## Lasso Regression

Lasso Regression is used to predict repository popularity while applying regularization to reduce the influence of less important features.

In [16]:
# Train Lasso Regression

lasso_model = models["Lasso"]

lasso_model.fit(
    X_train_sample,
    y_train_sample
)

print("Lasso Regression trained successfully.")

Lasso Regression trained successfully.


c:\Users\srich\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:784: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.973111e+09, tolerance: 2.885e+07
  model = cd_fast.sparse_enet_coordinate_descent(


In [17]:
# Make predictions using Lasso Regression

y_train_pred_lasso = lasso_model.predict(X_train_sample)
y_test_pred_lasso = lasso_model.predict(X_test_sample)

print("Lasso Regression predictions completed.")
print("Training predictions:", len(y_train_pred_lasso))
print("Testing predictions :", len(y_test_pred_lasso))

Lasso Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [18]:
# Evaluate Lasso Regression

train_rmse_lasso = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_lasso)
)

test_rmse_lasso = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_lasso)
)

train_r2_lasso = r2_score(
    y_train_sample,
    y_train_pred_lasso
)

test_r2_lasso = r2_score(
    y_test_sample,
    y_test_pred_lasso
)

print("Lasso Regression Evaluation")
print("---------------------------")
print("Train RMSE:", train_rmse_lasso)
print("Test RMSE :", test_rmse_lasso)
print("Train R²  :", train_r2_lasso)
print("Test R²   :", test_r2_lasso)

Lasso Regression Evaluation
---------------------------
Train RMSE: 77.51813043975584
Test RMSE : 2148.6377079442686
Train R²  : 0.9989585280418396
Test R²   : -0.07164943218231201


---

## Ridge Regression

Ridge Regression is used to predict repository popularity while applying L2 regularization to reduce overfitting in the high-dimensional feature space.

In [19]:
# Train Ridge Regression

ridge_model = models["Ridge"]

ridge_model.fit(
    X_train_sample,
    y_train_sample
)

print("Ridge Regression trained successfully.")

Ridge Regression trained successfully.


In [20]:
# Make predictions using Ridge Regression

y_train_pred_ridge = ridge_model.predict(X_train_sample)
y_test_pred_ridge = ridge_model.predict(X_test_sample)

print("Ridge Regression predictions completed.")
print("Training predictions:", len(y_train_pred_ridge))
print("Testing predictions :", len(y_test_pred_ridge))

Ridge Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [21]:
# Evaluate Ridge Regression

train_rmse_ridge = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_ridge)
)

test_rmse_ridge = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_ridge)
)

train_r2_ridge = r2_score(
    y_train_sample,
    y_train_pred_ridge
)

test_r2_ridge = r2_score(
    y_test_sample,
    y_test_pred_ridge
)

print("Ridge Regression Evaluation")
print("---------------------------")
print("Train RMSE:", train_rmse_ridge)
print("Test RMSE :", test_rmse_ridge)
print("Train R²  :", train_r2_ridge)
print("Test R²   :", test_r2_ridge)

Ridge Regression Evaluation
---------------------------
Train RMSE: 575.7668799088742
Test RMSE : 1725.2834549719648
Train R²  : 0.9425446391105652
Test R²   : 0.30904895067214966


---

## KNN Regression

KNN Regression predicts repository popularity by using the target values of the most similar repositories in the feature space.

In [22]:
# Train KNN Regression

knn_model = models["KNN"]

knn_model.fit(
    X_train_sample,
    y_train_sample
)

print("KNN Regression trained successfully.")

KNN Regression trained successfully.


In [23]:
# Make predictions using KNN Regression

y_train_pred_knn = knn_model.predict(X_train_sample)
y_test_pred_knn = knn_model.predict(X_test_sample)

print("KNN Regression predictions completed.")
print("Training predictions:", len(y_train_pred_knn))
print("Testing predictions :", len(y_test_pred_knn))

KNN Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [24]:
# Evaluate KNN Regression

train_rmse_knn = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_knn)
)

test_rmse_knn = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_knn)
)

train_r2_knn = r2_score(
    y_train_sample,
    y_train_pred_knn
)

test_r2_knn = r2_score(
    y_test_sample,
    y_test_pred_knn
)

print("KNN Regression Evaluation")
print("-------------------------")
print("Train RMSE:", train_rmse_knn)
print("Test RMSE :", test_rmse_knn)
print("Train R²  :", train_r2_knn)
print("Test R²   :", test_r2_knn)

KNN Regression Evaluation
-------------------------
Train RMSE: 1266.2949819478872
Test RMSE : 1288.4766067430173
Train R²  : 0.7220882729500915
Test R²   : 0.6146286854637213


---

## SVM Regression

SVM Regression is used to predict repository popularity by finding a regression function that provides the best fit while controlling model complexity.

In [25]:
# Train SVM Regression

svm_model = models["SVM"]

svm_model.fit(
    X_train_sample,
    y_train_sample
)

print("SVM Regression trained successfully.")

SVM Regression trained successfully.


In [26]:
# Make predictions using SVM Regression

y_train_pred_svm = svm_model.predict(X_train_sample)
y_test_pred_svm = svm_model.predict(X_test_sample)

print("SVM Regression predictions completed.")
print("Training predictions:", len(y_train_pred_svm))
print("Testing predictions :", len(y_test_pred_svm))

SVM Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [27]:
# Evaluate SVM Regression

train_rmse_svm = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_svm)
)

test_rmse_svm = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_svm)
)

train_r2_svm = r2_score(
    y_train_sample,
    y_train_pred_svm
)

test_r2_svm = r2_score(
    y_test_sample,
    y_test_pred_svm
)

print("SVM Regression Evaluation")
print("-------------------------")
print("Train RMSE:", train_rmse_svm)
print("Test RMSE :", test_rmse_svm)
print("Train R²  :", train_r2_svm)
print("Test R²   :", test_r2_svm)

SVM Regression Evaluation
-------------------------
Train RMSE: 1503.9406768425386
Test RMSE : 1669.211765226416
Train R²  : 0.6079887659802179
Test R²   : 0.35323101955693803


---

## Decision Tree Regression

Decision Tree Regression is used to model non-linear relationships between repository characteristics and repository popularity.

In [28]:
# Train Decision Tree Regression

dt_model = models["Decision Tree"]

dt_model.fit(
    X_train_sample,
    y_train_sample
)

print("Decision Tree Regression trained successfully.")

Decision Tree Regression trained successfully.


In [29]:
# Make predictions using Decision Tree Regression

y_train_pred_dt = dt_model.predict(X_train_sample)
y_test_pred_dt = dt_model.predict(X_test_sample)

print("Decision Tree Regression predictions completed.")
print("Training predictions:", len(y_train_pred_dt))
print("Testing predictions :", len(y_test_pred_dt))

Decision Tree Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [30]:
# Evaluate Decision Tree Regression

train_rmse_dt = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_dt)
)

test_rmse_dt = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_dt)
)

train_r2_dt = r2_score(
    y_train_sample,
    y_train_pred_dt
)

test_r2_dt = r2_score(
    y_test_sample,
    y_test_pred_dt
)

print("Decision Tree Regression Evaluation")
print("-----------------------------------")
print("Train RMSE:", train_rmse_dt)
print("Test RMSE :", test_rmse_dt)
print("Train R²  :", train_r2_dt)
print("Test R²   :", test_r2_dt)

Decision Tree Regression Evaluation
-----------------------------------
Train RMSE: 0.0
Test RMSE : 1494.3066129814188
Train R²  : 1.0
Test R²   : 0.4816707167537456


---

## Random Forest Regression

Random Forest Regression combines multiple decision trees to improve prediction stability and reduce the limitations of a single decision tree.

In [31]:
# Train Random Forest Regression

rf_model = models["Random Forest"]

rf_model.fit(
    X_train_sample,
    y_train_sample
)

print("Random Forest Regression trained successfully.")

Random Forest Regression trained successfully.


In [32]:
# Make predictions using Random Forest Regression

y_train_pred_rf = rf_model.predict(X_train_sample)
y_test_pred_rf = rf_model.predict(X_test_sample)

print("Random Forest Regression predictions completed.")
print("Training predictions:", len(y_train_pred_rf))
print("Testing predictions :", len(y_test_pred_rf))

Random Forest Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [33]:
# Evaluate Random Forest Regression

train_rmse_rf = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_rf)
)

test_rmse_rf = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_rf)
)

train_r2_rf = r2_score(
    y_train_sample,
    y_train_pred_rf
)

test_r2_rf = r2_score(
    y_test_sample,
    y_test_pred_rf
)

print("Random Forest Regression Evaluation")
print("-----------------------------------")
print("Train RMSE:", train_rmse_rf)
print("Test RMSE :", test_rmse_rf)
print("Train R²  :", train_r2_rf)
print("Test R²   :", test_r2_rf)

Random Forest Regression Evaluation
-----------------------------------
Train RMSE: 558.5601069523976
Test RMSE : 1306.9341786107477
Train R²  : 0.9459274400717942
Test R²   : 0.6035086292798619


---

## XGBoost Regression

XGBoost Regression is used to model complex non-linear relationships and predict repository popularity using an ensemble of boosted decision trees.

In [34]:
# Train XGBoost Regression

xgb_model = models["XGBoost"]

xgb_model.fit(
    X_train_sample,
    y_train_sample
)

print("XGBoost Regression trained successfully.")

XGBoost Regression trained successfully.


In [35]:
# Make predictions using XGBoost Regression

y_train_pred_xgb = xgb_model.predict(X_train_sample)
y_test_pred_xgb = xgb_model.predict(X_test_sample)

print("XGBoost Regression predictions completed.")
print("Training predictions:", len(y_train_pred_xgb))
print("Testing predictions :", len(y_test_pred_xgb))

XGBoost Regression predictions completed.
Training predictions: 50000
Testing predictions : 10000


In [36]:
# Evaluate XGBoost Regression

train_rmse_xgb = np.sqrt(
    mean_squared_error(y_train_sample, y_train_pred_xgb)
)

test_rmse_xgb = np.sqrt(
    mean_squared_error(y_test_sample, y_test_pred_xgb)
)

train_r2_xgb = r2_score(
    y_train_sample,
    y_train_pred_xgb
)

test_r2_xgb = r2_score(
    y_test_sample,
    y_test_pred_xgb
)

print("XGBoost Regression Evaluation")
print("----------------------------")
print("Train RMSE:", train_rmse_xgb)
print("Test RMSE :", test_rmse_xgb)
print("Train R²  :", train_r2_xgb)
print("Test R²   :", test_r2_xgb)

XGBoost Regression Evaluation
----------------------------
Train RMSE: 569.6138220320852
Test RMSE : 1739.5717863888228
Train R²  : 0.9437661170959473
Test R²   : 0.29755699634552


---

## Bias-Variance Trade-off

The difference between training and testing RMSE and R² scores is analyzed to determine whether each regression model is underfitting, overfitting, or providing a good fit.

In [37]:
# Create Bias-Variance comparison table

regression_results = pd.DataFrame({
    "Models": [
        "Linear Regression",
        "Polynomial Regression",
        "Lasso",
        "Ridge",
        "KNN",
        "SVM",
        "Decision Tree",
        "Random Forest",
        "XGBoost"
    ],
    "TrainLoss (RMSE)": [
        train_rmse_lr,
        train_rmse_poly,
        train_rmse_lasso,
        train_rmse_ridge,
        train_rmse_knn,
        train_rmse_svm,
        train_rmse_dt,
        train_rmse_rf,
        train_rmse_xgb
    ],
    "TestLoss (RMSE)": [
        test_rmse_lr,
        test_rmse_poly,
        test_rmse_lasso,
        test_rmse_ridge,
        test_rmse_knn,
        test_rmse_svm,
        test_rmse_dt,
        test_rmse_rf,
        test_rmse_xgb
    ],
    "TrainScore (R²)": [
        train_r2_lr,
        train_r2_poly,
        train_r2_lasso,
        train_r2_ridge,
        train_r2_knn,
        train_r2_svm,
        train_r2_dt,
        train_r2_rf,
        train_r2_xgb
    ],
    "TestScore (R²)": [
        test_r2_lr,
        test_r2_poly,
        test_r2_lasso,
        test_r2_ridge,
        test_r2_knn,
        test_r2_svm,
        test_r2_dt,
        test_r2_rf,
        test_r2_xgb
    ]
})

# Calculate the train-test R² gap
r2_gap = (
    regression_results["TrainScore (R²)"]
    - regression_results["TestScore (R²)"]
)

# Classify model fit
regression_results["Bias-Variance (Fit)"] = [
    "Goodfit" if gap <= 0.20 else "Overfit"
    for gap in r2_gap
]

print("Regression Bias-Variance Analysis")
print("---------------------------------")

display(
    regression_results.round(4)
)

Regression Bias-Variance Analysis
---------------------------------


,Models,TrainLoss (RMSE),TestLoss (RMSE),TrainScore (R²),TestScore (R²),Bias-Variance (Fit)
0,Linear Regression,7.8609,1840.6021,1.0000,0.2136,Overfit
1,Polynomial Regression,1037.0034,1784.7246,0.8136,0.2606,Overfit
2,Lasso,77.5181,2148.6377,0.9990,-0.0716,Overfit
3,Ridge,575.7669,1725.2835,0.9425,0.3090,Overfit
4,KNN,1266.2950,1288.4766,0.7221,0.6146,Goodfit
5,SVM,1503.9407,1669.2118,0.6080,0.3532,Overfit
6,Decision Tree,0.0000,1494.3066,1.0000,0.4817,Overfit
7,Random Forest,558.5601,1306.9342,0.9459,0.6035,Overfit
8,XGBoost,569.6138,1739.5718,0.9438,0.2976,Overfit


---

## Cross-Validation

Cross-validation was considered for evaluating the generalization performance of the regression models.

However, the P2 feature matrix contains **416,751 features**, and computationally expensive models such as Random Forest require substantial training time even on the model-study sample.

Therefore, to keep the workflow computationally practical, model comparison is based on:

- Independent test-set RMSE and R²
- Train-test performance comparison
- Bias-Variance analysis

The independent test set provides an unseen dataset for evaluating how well each regression model generalizes to new repository data.

---

## Final Model Selection

Based on the model comparison and bias-variance analysis, KNN Regression is selected as the final model for Repository Popularity Prediction.

- KNN achieved the lowest Test RMSE of **1288.48** among the evaluated regression models.
- KNN achieved the highest Test R² Score of **0.6146**.
- The relatively small difference between its training and testing performance indicates good generalization.
- Therefore, KNN Regression is selected as the final model for repository popularity prediction.

In [38]:
# Train Final KNN Regression Model on Full Training Data

knn_final = KNeighborsRegressor(
    n_neighbors=5,
    n_jobs=-1
)

knn_final.fit(
    X_train,
    y_train
)

print("Final KNN Regression model trained successfully.")
print("Training rows:", X_train.shape[0])
print("Training features:", X_train.shape[1])

Final KNN Regression model trained successfully.
Training rows: 261438
Training features: 416751


In [39]:
# Make Final KNN Predictions

y_train_pred_final = knn_final.predict(X_train)
y_test_pred_final = knn_final.predict(X_test)

print("Final KNN predictions completed.")
print("Training predictions:", len(y_train_pred_final))
print("Testing predictions :", len(y_test_pred_final))

Final KNN predictions completed.
Training predictions: 261438
Testing predictions : 65360


In [40]:
# Evaluate Final KNN Regression Model

from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

final_train_rmse = np.sqrt(
    mean_squared_error(y_train, y_train_pred_final)
)

final_test_rmse = np.sqrt(
    mean_squared_error(y_test, y_test_pred_final)
)

final_train_r2 = r2_score(
    y_train,
    y_train_pred_final
)

final_test_r2 = r2_score(
    y_test,
    y_test_pred_final
)

print("Final KNN Regression Evaluation")
print("--------------------------------")
print("Train RMSE:", final_train_rmse)
print("Test RMSE :", final_test_rmse)
print("Train R²  :", final_train_r2)
print("Test R²   :", final_test_r2)

Final KNN Regression Evaluation
--------------------------------
Train RMSE: 935.1937813594875
Test RMSE : 1195.5153060316445
Train R²  : 0.7890588983896218
Test R²   : 0.5226717440938935


----

# Hyperparameter Tuning

Hyperparameter tuning was considered to further improve the performance of the selected KNN Regression model.

- KNN was selected as the final regression model based on the model-study results.
- The selected KNN configuration uses **n_neighbors = 5**.
- The final KNN model was trained using the complete training dataset of **261,438 rows**.
- The final model achieved a Test RMSE of **1195.52** and Test R² of **0.5227** on the complete unseen test dataset.
- The P2 feature matrix contains **416,751 features**, making extensive hyperparameter searches computationally expensive.
- Therefore, additional hyperparameter tuning was not performed, and the existing KNN configuration is retained as the final model configuration.

---

## Saving Model & Real-Time Prediction

The trained, evaluated, and selected KNN Regression model is saved for future use. The saved model can be loaded later and used to predict repository popularity for unseen repository data using the same preprocessing applied during training.

In [41]:
# Save Final KNN Model

import joblib

joblib.dump(
    knn_final,
    "../models/knn_final.pkl"
)

print("Final KNN Regression model saved successfully.")

Final KNN Regression model saved successfully.


In [42]:
# Load Saved Final KNN Model

import joblib

knn_loaded = joblib.load(
    "../models/knn_final.pkl"
)

print("Final KNN Regression model loaded successfully.")
print("Model:", knn_loaded)

Final KNN Regression model loaded successfully.
Model: KNeighborsRegressor(n_jobs=-1)


---

## Real-Time Prediction

The saved KNN Regression model is loaded and used to predict the expected number of stars for an unseen repository.

The new repository input is transformed using the same feature engineering and preprocessing steps used during model training before generating the prediction.

In [43]:
# Create an unseen repository input for real-time prediction

new_repository = pd.DataFrame([{
    "license": "MIT",
    "is_forked": False,
    "language": "Python",
    "forks": 25,
    "repository_name_length": 12,
    "repository_owner": "example-user",
    "repository_name": "ai-project"
}])

print("Unseen repository input:")
display(new_repository)

Unseen repository input:


,license,is_forked,language,forks,repository_name_length,repository_owner,repository_name
0,MIT,False,Python,25,12,example-user,ai-project


In [44]:
# Load P2 preprocessing artifacts

import joblib

p2_encoder = joblib.load("../models/p2_encoder.pkl")
p2_imputer = joblib.load("../models/p2_imputer.pkl")
p2_scaler = joblib.load("../models/p2_scaler.pkl")

print("P2 preprocessing artifacts loaded successfully.")

P2 preprocessing artifacts loaded successfully.


In [45]:
# Transform unseen repository using the saved P2 preprocessing pipeline

# Separate categorical and numerical features
categorical_features = new_repository[
    [
        "license",
        "language",
        "is_forked",
        "repository_owner",
        "repository_name"
    ]
].copy()

numerical_features = new_repository[
    [
        "forks",
        "repository_name_length"
    ]
].copy()

# Match the representation used during P2 feature engineering
categorical_features["is_forked"] = categorical_features["is_forked"].map({
    False: "0.0",
    True: "1.0"
})

# Encode categorical features
X_new_encoded = p2_encoder.transform(
    categorical_features
)

# Impute and scale numerical features
X_new_imputed = p2_imputer.transform(
    numerical_features
)

X_new_scaled = p2_scaler.transform(
    X_new_imputed
)

print("Categorical features:", X_new_encoded.shape)
print("Numerical features  :", X_new_scaled.shape)

Categorical features: (1, 416749)
Numerical features  : (1, 2)


In [46]:
# Build final feature matrix for the unseen repository

from scipy.sparse import hstack, csr_matrix

X_new = hstack([
    X_new_encoded,
    csr_matrix(X_new_scaled)
]).tocsr().astype("float32")

print("Final unseen feature matrix:", X_new.shape)

Final unseen feature matrix: (1, 416751)


In [47]:
# Predict repository popularity using the loaded KNN model

predicted_stars = knn_loaded.predict(X_new)[0]

print("Predicted Stars:", predicted_stars)

Predicted Stars: 300.4


## Regression Conclusion

The regression models were evaluated using RMSE, R² Score, and Bias-Variance analysis.

- KNN Regression was selected as the final model based on the lowest Test RMSE and highest Test R² among the evaluated models.
- The final KNN model achieved a Test RMSE of **1195.52** and Test R² of **0.5227** on the complete unseen test dataset.
- The final KNN Regression model was saved and successfully loaded for future predictions.
- Real-time prediction was successfully demonstrated on an unseen repository, where the model predicted approximately **300 stars**.

---